File: src/api/roster_intelligence.py
"""Assembly for ``GET /api/roster/intelligence``.

The transport-facing composition of the canonical roster-intelligence
chain — meaningful core → Team Strength → Team Weakness → age/value
portfolio — for one league, with the requesting team broken out.

**This module computes nothing.**  Every number comes from
``src/roster_intel/{core,strength,weakness,age_portfolio}.py``; this is
the shell that loads inputs and shapes a response, the same relationship
``server.py::get_gameplan`` has with ``src/api/gameplan.py``.

The roster source is the CONTRACT, and that is load-bearing
===========================================================

Rosters come from ``data_contract.contract_roster_pools`` — the same
builder the ``optimalLineup`` stamp uses — not from the ROS
team-strength snapshot.  Two defects made the snapshot the wrong
source, and both are gone as a consequence rather than as a patch:

**It carried the wrong QUANTITY.**  Its ``rosValue`` is "a normalized
log-rank index on 0-100, not points, and not projection-aware"
(``src/ros/aggregate.py``) — a rest-of-season PRODUCTION measure.
``MASTER_PRODUCT_PLAN`` §4.1 says Team Strength "is not Power Ranking,
Playoff Odds, or ROS production" and must use canonical league values.
Team Strength was therefore summing the wrong number entirely; it now
sums ``rankDerivedValue`` on the canonical 1-9999 dynasty scale.

**It flattened unpriced players into real zeros before anyone could
see them.**  ``ros/team_strength.py`` appends every row it cannot match
with ``ros_value=0.0`` — its own comment calls that "a real
missing-is-zero coercion … left in place DELIBERATELY" — so
``unpriced_ids`` was structurally empty: not because every player was
priced, but because the unpriced ones arrived indistinguishable from
players genuinely worth nothing.  V1's MISSING IS NEVER ZERO rule
requires a consumer to tell those two apart, and a source that erases
the distinction makes that impossible downstream however correct the
consumer is.

*(Corrected 2026-08-19, integration review.  This said the writer
DROPPED those rows.  It does not.  The consequence for a consumer is
the same and the repair here is unchanged, but a coercion and a
deletion are fixed differently at the source — and §11 of the lane doc
hands the integration lane an instruction premised on the description,
so it had to be right.)*

Measured on the contract built from ``dynasty_export_20260818_191848.zip``
via ``build_api_data_contract``: **83 of 660 rostered players (12.6%)
carry no canonical value**, and every one is now reported in
``unpricedIds`` instead of reading as zero.

A third consequence, worth stating because it removes a real
operational fragility: the endpoint no longer needs the ROS refresh to
have run at all.  The contract is always loaded, so roster
intelligence cannot be dark because a different lane's timer failed.

**Positional ranks are measured against the contract board.**  "Top 12
QB" is a statement about a player's standing, not about who owns him,
so the population is every priced non-pick row on the canonical board —
and it is stamped, because a consumer comparing two teams must know
they were measured the same way.
"""

from __future__ import annotations

import math
import time
from typing import Any, Mapping

from src.api.data_contract import contract_roster_pools
from src.roster_intel.age_portfolio import (
    build_age_portfolio,
    build_youth_curve,
    rank_age_portfolios,
)
from src.roster_intel.core import build_meaningful_core
from src.roster_intel.droppability import league_droppability, team_droppability
from src.roster_intel.exposure import build_nfl_exposure, exposure_from_core
from src.roster_intel.strength import build_team_strength, rank_team_strengths
from src.roster_intel.weakness import build_position_ranks, build_team_weakness

__all__ = [
    "ROSTER_INTELLIGENCE_CONTRACT_VERSION",
    "TeamNotInLeague",
    "build_league_roster_intelligence",
    "get_team_roster_intelligence",
]

ROSTER_INTELLIGENCE_CONTRACT_VERSION = "roster-intelligence/2026-08-18.v1"

#: Where rosters and values come from.  Named in the payload because a
#: consumer comparing two responses must know they were built the same
#: way — and because this value CHANGED (it was the ROS snapshot, which
#: carried a production index and pre-deleted unpriced players).
_ROSTER_SOURCE = "canonical_contract"
_UNPRICED_VISIBILITY = (
    "rostered players the canonical board did not price are reported in "
    "unpricedIds and excluded from every value aggregate; they are never "
    "counted as zero"
)
_RANK_POPULATION = "contract_board_priced_players"


class TeamNotInLeague(Exception):
    """The requested owner has no roster in this league."""


def _board_players(contract: Mapping[str, Any] | None) -> list[tuple[str, str, float | None]]:
    """``(playerKey, position, value)`` for every non-pick board row.

    ``playerKey`` is the CANONICAL NAME, not the Sleeper id, because
    that is what ``contract_roster_pools`` keys a ``RosterPlayer`` by —
    and every consumer of this list looks its rows up by
    ``CoreMember.player_id``.

    That is not a stylistic choice; keying it by ``playerId`` is a
    silent-failure bug, and this function had it. Positional ranks then
    matched nothing, so every weakness rung reported UNKNOWN, and the
    youth curve was empty, so every Young Core Index came back ``None``.
    Nothing raised — the payload was fully shaped and entirely
    unpopulated, which is precisely the failure mode the missing-vs-zero
    discipline exists to make visible. It was caught by running the real
    12-team board and seeing twelve ``None`` indices, not by a test.

    Picks are excluded because they are not players and cannot hold a
    positional rank; unpriced rows are carried through with ``None`` so
    ``build_position_ranks`` excludes them for its own stated reason
    rather than never seeing them.
    """
    rows: list[tuple[str, str, float | None]] = []
    if not isinstance(contract, Mapping):
        return rows
    for row in contract.get("playersArray") or []:
        if not isinstance(row, Mapping) or row.get("assetClass") == "pick":
            continue
        position = str(row.get("position") or "").strip()
        if not position:
            continue
        value = row.get("rankDerivedValue")
        priced = float(value) if isinstance(value, (int, float)) else None
        for key in (row.get("canonicalName"), row.get("displayName")):
            if key:
                rows.append((str(key), position, priced))
                break
    return rows


def _ages(contract: Mapping[str, Any] | None) -> dict[str, float | None]:
    """``{playerName: age}`` from the canonical board.

    Keyed by the SAME name ``contract_roster_pools`` keys players by, so
    an age lookup cannot silently miss every player because two halves
    of this module disagreed about the join key.  A non-positive age is
    dropped rather than carried: Sleeper writes ``0`` for an unresolved
    record, and ``0`` is exactly the value that would make a roster look
    historically young.
    """
    out: dict[str, float | None] = {}
    if not isinstance(contract, Mapping):
        return out
    for row in contract.get("playersArray") or []:
        if not isinstance(row, Mapping) or row.get("assetClass") == "pick":
            continue
        age = row.get("age")
        if not isinstance(age, (int, float)) or age <= 0:
            continue
        for key in (row.get("canonicalName"), row.get("displayName")):
            if key:
                out.setdefault(str(key), float(age))
    return out


def _nfl_teams(contract: Mapping[str, Any] | None) -> dict[str, str]:
    """``{playerName: nflTeam}`` from the canonical board.

    Keyed the same way ``contract_roster_pools`` keys players, for the reason
    ``_board_players`` documents at length: a join key that disagrees with the
    pools fails silently and completely.

    An empty ``team`` is left OUT rather than stored as ``""`` — the exposure
    owner reads absence as UNKNOWN and reports the player, and an empty string
    would instead create a bucket that holds a share.  Measured on the live
    board, 0 of 660 rostered players lack one, but 25 carry ``FA``, which is a
    real answer (unsigned) and not a missing one.
    """
    out: dict[str, str] = {}
    if not isinstance(contract, Mapping):
        return out
    for row in contract.get("playersArray") or []:
        if not isinstance(row, Mapping) or row.get("assetClass") == "pick":
            continue
        team = str(row.get("team") or "").strip()
        if not team:
            continue
        for key in (row.get("canonicalName"), row.get("displayName")):
            if key:
                out.setdefault(str(key), team)
    return out


def build_league_roster_intelligence(
    contract: Mapping[str, Any] | None,
    *,
    team_count: int | None = None,
    include_droppability: bool = False,
) -> dict[str, Any]:
    """Core / strength / weakness / age for EVERY team in one league.

    League-wide because three of the four outputs are league-RELATIVE:
    Team Strength rank, the Young Core Index and each room's percentile
    are all undefined for a single team.  Computing one team in
    isolation and then inventing its rank is the failure this shape
    prevents.
    """
    t0 = time.perf_counter()
    pools, slots, slot_source = contract_roster_pools(dict(contract or {}))
    team_names = _team_names(contract)
    ages = _ages(contract)

    board = _board_players(contract)
    ranks = build_position_ranks(board, population=_RANK_POPULATION)
    youth = build_youth_curve([(position, ages.get(pid)) for pid, position, _ in board])

    # The league's OWN size, with the registry's declared count preferred
    # when supplied: a contract missing one roster must not shrink every
    # weakness threshold.
    n_teams = int(team_count) if team_count else len(pools)

    cores = {oid: build_meaningful_core(pool, slots) for oid, pool in pools.items()}
    strengths = rank_team_strengths({oid: build_team_strength(core) for oid, core in cores.items()})
    weaknesses = {
        oid: build_team_weakness(core, ranks, team_count=n_teams) for oid, core in cores.items()
    }
    portfolios = rank_age_portfolios(
        {
            oid: build_age_portfolio(
                core,
                ages,
                youth=youth,
                full_roster=[
                    (pl.player_id, float(pl.ros_value))
                    for pl in pools[oid]
                    if pl.ros_value is not None
                ],
            )
            for oid, core in cores.items()
        }
    )

    drops = league_droppability(contract) if include_droppability else {}
    nfl_teams = _nfl_teams(contract)
    roster_values = {pl.player_id: pl.ros_value for pool in pools.values() for pl in pool}

    teams = {
        oid: {
            "ownerId": oid,
            "teamName": team_names.get(oid, ""),
            "rosteredCount": len(pools[oid]),
            "core": cores[oid].to_dict(),
            "strength": strengths[oid].to_dict(),
            "weakness": weaknesses[oid].to_dict(),
            "agePortfolio": portfolios[oid].to_dict(),
            # Descriptive only (C2-EXP-01).  Two scopes, named separately,
            # because "how concentrated is what plays" and "how concentrated
            # is the capital" are different questions with different answers.
            "nflExposure": {
                "core": exposure_from_core(cores[oid], teams=nfl_teams).to_dict(),
                "fullRoster": build_nfl_exposure(
                    [pl.player_id for pl in pools[oid]],
                    teams=nfl_teams,
                    values=roster_values,
                    positions={pl.player_id: pl.position for pl in pools[oid]},
                    scope="full_roster",
                ).to_dict(),
            },
            **({"droppability": drops[oid]} if oid in drops else {}),
        }
        for oid in pools
    }

    return {
        "contractVersion": ROSTER_INTELLIGENCE_CONTRACT_VERSION,
        "leagueKey": (contract or {}).get("meta", {}).get("leagueKey"),
        "teamCount": n_teams,
        "starterSlots": slots,
        "slotSource": slot_source,
        "rosterSource": _ROSTER_SOURCE,
        "unpricedVisibility": _UNPRICED_VISIBILITY,
        "rankPopulation": _RANK_POPULATION,
        # Stamped whether or not it was asked for: "you did not request
        # droppability" and "this team has nothing droppable" must not
        # read the same, and an absent key alone cannot tell them apart.
        "droppabilityIncluded": bool(include_droppability),
        "teams": teams,
        "timing": {"computeMs": round((time.perf_counter() - t0) * 1000.0, 1)},
    }


def _team_names(contract: Mapping[str, Any] | None) -> dict[str, str]:
    """``{ownerId: teamName}`` from the contract's sleeper block."""
    out: dict[str, str] = {}
    sleeper = (contract or {}).get("sleeper")
    if not isinstance(sleeper, Mapping):
        return out
    for team in sleeper.get("teams") or []:
        if isinstance(team, Mapping):
            oid = str(team.get("ownerId") or "")
            if oid:
                out[oid] = str(team.get("name") or team.get("sleeperTeamName") or "")
    return out


def _league_context_order(item: tuple[str, Mapping[str, Any]]) -> tuple[bool, float, str]:
    """Sort key for the league context: ranked teams first, best first.

    An unranked team sorts to the END on an explicit ``inf``, never on a
    coerced ``0``. The coercion is what the decision-coercion gate flags
    and it is right to: ``rank or 0`` reads an ABSENT rank as the best
    possible one, so the leading ``is None`` element is the only thing
    keeping such a team out of first place. Two guards where one should
    do it, and the fragile one is invisible.

    ``inf`` needs no companion guard — it is last on its own — so the
    ``is None`` element survives only to keep every unranked team
    grouped, and ``owner_id`` makes ties deterministic.
    """
    owner_id, team = item
    rank = team["strength"]["leagueRank"]
    return (rank is None, math.inf if rank is None else float(rank), owner_id)


def get_team_roster_intelligence(
    contract: Mapping[str, Any] | None,
    owner_id: str,
    *,
    team_count: int | None = None,
    include_droppability: bool = False,
) -> dict[str, Any]:
    """One team's roster intelligence, plus the league context it is
    ranked against.

    Raises :class:`TeamNotInLeague` when the owner has no roster in this
    contract — distinct from "this team has nothing", which is a real
    answer with a real payload.

    ``league_key`` / ``scoring_profile`` were parameters while this read
    the ROS snapshot through ``gameplan.get_league_bundle``; the
    contract carries its own league identity, so they are gone rather
    than left accepted-and-ignored.
    """
    league = build_league_roster_intelligence(contract, team_count=team_count)
    team = league["teams"].get(str(owner_id))
    if team is None:
        raise TeamNotInLeague(str(owner_id))

    if include_droppability:
        # ONE team's ladder, not the league's — the league loop costs 13x
        # this and the other eleven ladders are not part of this answer.
        team = dict(team)
        team["droppability"] = team_droppability(contract, owner_id=str(owner_id))

    payload = dict(league)
    payload["droppabilityIncluded"] = bool(include_droppability)
    payload["team"] = team
    # League-relative context without shipping every roster twice: the
    # ranks a consumer needs to place this team, and nothing more.
    payload["leagueContext"] = [
        {
            "ownerId": oid,
            "teamName": t["teamName"],
            "strengthTotal": t["strength"]["total"],
            "strengthRank": t["strength"]["leagueRank"],
            "youngCoreIndex": t["agePortfolio"]["youngCoreIndex"],
            "valueWeightedCoreAge": t["agePortfolio"]["valueWeightedCoreAge"],
        }
        for oid, t in sorted(league["teams"].items(), key=_league_context_order)
    ]
    payload.pop("teams", None)
    return payload

File: frontend/lib/roster-intelligence.js
/**
 * roster-intelligence.js — pure display helpers for the canonical
 * Team Strength surface served by `GET /api/roster/intelligence`.
 *
 * The canonical owner is `src/roster_intel/strength.py` (feature
 * inventory row 1.1), reached through `src/api/roster_intelligence.py`.
 * Every number this module hands to a component was computed there.
 * Nothing here sums player values, weights anything, or classifies a
 * team — the whole point of the migration that introduced this file is
 * that `/rosters` stopped doing exactly that.
 *
 * The rule it exists to keep, from CLAUDE.md: "There is no frontend
 * ranking engine, period — not even a fallback." A materializer may
 * reshape and format a backend number; it may not produce one.
 *
 * ── Three quantities, deliberately not merged ─────────────────────────
 * `V1-35` says the named team quantities "may not collapse into one
 * team score", and the backend already publishes them side by side for
 * that reason. This module keeps them apart in the vocabulary it hands
 * the UI:
 *
 *   `total`            THE Team Strength number — the meaningful core
 *   `fullRosterValue`  the whole-roster PORTFOLIO sum, named separately
 *                      because a 58-man best-ball roster books bench
 *                      player #40 at full market value
 *   `starterValue` /   the split inside the core; a diagnostic about
 *   `reserveValue`     lineup structure, not a second strength
 *
 * ── Missing is never zero ─────────────────────────────────────────────
 * Four absences are carried through as absences rather than numbers,
 * because the backend went to some trouble to distinguish them:
 *
 *   `available: false`   the lineup could not be read. NOT a strength
 *                        of 0 — see `build_team_strength`'s refusal
 *                        branch, which propagates the core's reason.
 *   `leagueRank: null`   NOT MEASURED, never "last". `rank_team_strengths`
 *                        excludes unreadable rosters from the ranking
 *                        population instead of ranking them bottom.
 *   `unpricedCount > 0`  part of the roster has no canonical value, so
 *                        the total is a statement about the rest.
 *   `unfilledStarterSlots`  the core could not fill the league's lineup,
 *                        so part of the strength is UNMEASURED.
 */

/** Rendered wherever a rank exists as a concept but was not measured. */
export const NOT_MEASURED = "Not measured";

/**
 * Classify a non-2xx `/api/roster/intelligence` response into the states
 * the page renders distinctly.
 *
 * Returns null for 2xx; otherwise `{ kind, message }` with kind one of
 * "not_ready" | "team_required" | "team_not_found" | "league" | "auth" |
 * "unavailable" | "error".
 *
 * `team_required` and `team_not_found` are separated from the generic
 * error on purpose: both are fixable by the user (pick a team / pick a
 * different one) and neither means the engine is broken.
 */
export function classifyRosterIntelligenceFailure(status, body) {
  if (status >= 200 && status < 300) return null;
  const code = body && typeof body === "object" ? body.error : "";
  const message =
    (body && typeof body === "object" && (body.message || body.detail)) || "";
  if (status === 401) {
    return { kind: "auth", message: "Sign in to see roster intelligence." };
  }
  if (code === "team_required") return { kind: "team_required", message };
  if (code === "team_not_found") return { kind: "team_not_found", message };
  if (code === "unknown_league" || code === "inactive_league" || code === "no_leagues_configured") {
    return { kind: "league", message };
  }
  if (code === "data_not_ready") return { kind: "not_ready", message };
  if (code === "roster_intelligence_unavailable") {
    return { kind: "unavailable", message };
  }
  return { kind: "error", message: message || `HTTP ${status}` };
}

/** A finite number, or null. Never a coerced 0 — that is the defect this
 *  whole surface exists to remove. */
function num(v) {
  return typeof v === "number" && Number.isFinite(v) ? v : null;
}

/**
 * The league's Team Strength ladder, from the payload's own
 * `leagueContext`.
 *
 * The backend already ordered it (`_league_context_order`: ranked teams
 * first, best first, unranked last on an explicit `inf`). We preserve
 * that order rather than re-sorting — re-sorting here would be a second
 * opinion about the ladder, and a `?? 0` in the comparator is precisely
 * how an unranked team becomes #1. Rows are stamped `measured` so a
 * consumer can render the absence rather than infer it from a null.
 */
export function teamStrengthLadder(payload, { myOwnerId = "" } = {}) {
  const context = Array.isArray(payload?.leagueContext) ? payload.leagueContext : [];
  return context.map((t) => {
    const rank = num(t?.strengthRank);
    return {
      ownerId: String(t?.ownerId ?? ""),
      teamName: String(t?.teamName ?? ""),
      rank,
      measured: rank !== null,
      rankLabel: rank === null ? NOT_MEASURED : `#${rank}`,
      strengthTotal: num(t?.strengthTotal),
      youngCoreIndex: num(t?.youngCoreIndex),
      valueWeightedCoreAge: num(t?.valueWeightedCoreAge),
      isMe: Boolean(myOwnerId) && String(t?.ownerId ?? "") === String(myOwnerId),
    };
  });
}

/**
 * The requested team's canonical strength block, reshaped for render.
 *
 * Returns null when the payload carries no team — distinct from a team
 * whose strength is `available: false`, which is a real answer with a
 * real reason and is returned with `available: false`.
 */
export function teamStrengthDetail(payload) {
  const team = payload?.team;
  const strength = team?.strength;
  if (!strength || typeof strength !== "object") return null;

  const order = Array.isArray(strength.positionOrder) ? strength.positionOrder : [];
  const byPosition = Array.isArray(strength.byPosition) ? strength.byPosition : [];
  // The backend declares the display order and the groups together.
  // We render the groups it published, in the order it published them —
  // no local group list, so a league running an exotic slot cannot lose
  // a column and FLEX cannot become one (a FLEX-seated RB is summed
  // under RB by `MeaningfulCore.by_position`).
  const positionIndex = new Map(byPosition.map((p) => [String(p?.position ?? ""), p]));
  const positions = (order.length ? order : byPosition.map((p) => p?.position))
    .map((name) => positionIndex.get(String(name)))
    .filter(Boolean)
    .map((p) => {
      const rank = num(p?.leagueRank);
      return {
        position: String(p?.position ?? ""),
        value: num(p?.value),
        count: num(p?.count) ?? 0,
        starterValue: num(p?.starterValue),
        starterCount: num(p?.starterCount) ?? 0,
        reserveValue: num(p?.reserveValue),
        reserveCount: num(p?.reserveCount) ?? 0,
        rank,
        rankLabel: rank === null ? NOT_MEASURED : `#${rank}`,
      };
    });

  const unfilledStarterSlots = Array.isArray(strength.unfilledStarterSlots)
    ? strength.unfilledStarterSlots.map(String)
    : [];
  const unfilledReserveSlots = Array.isArray(strength.unfilledReserveSlots)
    ? strength.unfilledReserveSlots.map(String)
    : [];
  const rank = num(strength.leagueRank);

  return {
    ownerId: String(team?.ownerId ?? ""),
    teamName: String(team?.teamName ?? ""),
    rosteredCount: num(team?.rosteredCount),
    available: strength.available !== false,
    unavailableReason: strength.unavailableReason || "",
    // Read verbatim. Not rounded here either: the backend already
    // rounded to 3dp and a second rounding rule is a second opinion.
    total: num(strength.total),
    starterValue: num(strength.starterValue),
    reserveValue: num(strength.reserveValue),
    // Named separately from `total` everywhere it is shown.
    fullRosterValue: num(strength.fullRosterValue),
    positions,
    unpricedCount: num(strength.unpricedCount) ?? 0,
    unfilledStarterSlots,
    unfilledReserveSlots,
    isComplete: strength.isComplete === true,
    rank,
    rankLabel: rank === null ? NOT_MEASURED : `#${rank}`,
    percentile: num(strength.leaguePercentile),
  };
}

/**
 * Why a total is only part of the story, as a list of plain sentences.
 *
 * Empty when the strength is complete. Derived from the backend's own
 * completeness fields; it invents no threshold and no severity.
 */
export function strengthCaveats(detail) {
  if (!detail || !detail.available) return [];
  const out = [];
  if (detail.unfilledStarterSlots.length) {
    out.push(
      `${detail.unfilledStarterSlots.length} starting ${
        detail.unfilledStarterSlots.length === 1 ? "slot" : "slots"
      } could not be filled (${detail.unfilledStarterSlots.join(", ")}), so part of ` +
        "this team's strength is unmeasured rather than low.",
    );
  }
  if (detail.unfilledReserveSlots.length) {
    out.push(
      `${detail.unfilledReserveSlots.length} reserve ${
        detail.unfilledReserveSlots.length === 1 ? "slot" : "slots"
      } could not be filled (${detail.unfilledReserveSlots.join(", ")}).`,
    );
  }
  if (detail.unpricedCount > 0) {
    out.push(
      `${detail.unpricedCount} rostered ${
        detail.unpricedCount === 1 ? "player has" : "players have"
      } no canonical value, so ${
        detail.unpricedCount === 1 ? "it is" : "they are"
      } excluded from the total rather than counted as zero.`,
    );
  }
  return out;
}

/** The Sleeper ownerId for a team NAME, which is what `settings.selectedTeam`
 *  stores. Returns "" when the name is unknown — the endpoint then falls back
 *  to the session's own team, which is the correct default. */
export function ownerIdForTeamName(sleeperTeams, teamName) {
  if (!teamName) return "";
  const match = (sleeperTeams || []).find((t) => t?.name === teamName);
  return match?.ownerId ? String(match.ownerId) : "";
}

/** Whole-number display for an uncapped aggregate. Aggregates may exceed
 *  9999 (inventory row 7.5) and are never clamped here. */
export function formatStrengthValue(value) {
  if (value === null || value === undefined) return "—";
  return Math.round(value).toLocaleString();
}
